sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -3320,8 +3320,27 @@ object GpuOverrides extends Logging {
       "hive hash operator",
       ExprChecks.projectOnly(TypeSig.INT, TypeSig.INT,
         repeatingParamCheck = Some(RepeatingParamCheck("input",
-          TypeSig.commonCudfTypes + TypeSig.NULL, TypeSig.all))),
+          (TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.STRUCT + TypeSig.ARRAY).nested() +
+              TypeSig.psNote(TypeEnum.ARRAY, "Nested levels exceeding 8 layers are not supported") +
+              TypeSig.psNote(TypeEnum.STRUCT, "Nested levels exceeding 8 layers are not supported"),
+          TypeSig.all))),
       (a, conf, p, r) => new ExprMeta[HiveHash](a, conf, p, r) {
+        override def tagExprForGpu(): Unit = {
+          def getMaxNestedDepth(inputType: DataType): Int = {
+            inputType match {
+              case at: ArrayType => 1 + getMaxNestedDepth(at.elementType)
+              case st: StructType =>
+                1 + st.map(f => getMaxNestedDepth(f.dataType)).max
+              case _ => 0 // primitive types
+            }
+          }
+          val maxDepth = a.children.map(c => getMaxNestedDepth(c.dataType)).max
+          if (maxDepth > 8) {
+            willNotWorkOnGpu(s"GPU HiveHash supports 8 levels at most for " +
```

Comment:
we'd better to have a test for this fallback?

Reply:
Nit: We should better retrieve this `maxDepth` from JNI instead, so if we change the max depth support in the future, we won't have to modify this code.

Check out this: https://github.com/NVIDIA/spark-rapids/blob/730ff2c64b67a2b5edd2552d97484a92ac1110a9/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGetJsonObject.scala#L105-L106